MacOS SDRPlay API Error While Resolving Native Library Lookup

**sdrtrunk Version**
master

**Describe the bug**
See application log snippet.

**Application Log**
```
2024-08-31 17:08:25.432 INFO  i.g.d.log.ApplicationLog -  Recordings:       /Users/sdr/SDRTrunk/recordings  [13MB/40MB 33%]
2024-08-31 17:08:25.452 INFO  i.g.dsheirer.gui.SDRTrunk - SDRPlay API native library preemptively loaded  [14MB/40MB 35%]
2024-08-31 17:08:25.526 INFO  i.g.d.util.ThreadPool - Application thread pool created SCHEDULED and CACHED executors threads  [17MB/40MB 43%]
2024-08-31 17:08:25.527 INFO  i.g.d.p.SystemProperties - SystemProperties - loaded [/Users/sdr/SDRTrunk/SDRTrunk.properties]  [17MB/40MB 43%]
2024-08-31 17:08:25.601 INFO  i.g.d.s.t.m.TunerManager - Discovering tuners ...  [21MB/40MB 54%]
2024-08-31 17:08:25.606 INFO  i.g.d.s.t.m.TunerManager - LibUsb API Version: 1.0.264  [21MB/40MB 54%]
2024-08-31 17:08:25.606 INFO  i.g.d.s.t.m.TunerManager - LibUsb Version: 1.0.24.11584  [21MB/40MB 54%]
2024-08-31 17:08:25.606 INFO  i.g.d.s.t.m.TunerManager - LibUsb - discovered [0] potential usb devices  [21MB/40MB 54%]
2024-08-31 17:08:25.607 INFO  i.g.d.s.t.m.TunerManager - LibUsb - starting HotPlug detection  [21MB/40MB 54%]
Exception in thread "main" java.lang.ExceptionInInitializerError
	at io.github.dsheirer.source.tuner.sdrplay.api.v3_07.sdrplay_api_h.sdrplay_api_Open(sdrplay_api_h.java:1550)
	at io.github.dsheirer.source.tuner.sdrplay.api.SDRplay.open(SDRplay.java:502)
	at io.github.dsheirer.source.tuner.sdrplay.api.SDRplay.<init>(SDRplay.java:98)
	at io.github.dsheirer.source.tuner.manager.TunerManager.discoverSdrPlayTuners(TunerManager.java:336)
	at io.github.dsheirer.source.tuner.manager.TunerManager.start(TunerManager.java:139)
	at io.github.dsheirer.gui.SDRTrunk.<init>(SDRTrunk.java:193)
	at io.github.dsheirer.gui.SDRTrunk.main(SDRTrunk.java:937)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.foreign.SymbolLookup.find(String)" because "io.github.dsheirer.source.tuner.sdrplay.api.v3_07.sdrplay_api_h.SYMBOL_LOOKUP" is null
	at io.github.dsheirer.source.tuner.sdrplay.api.v3_07.sdrplay_api_h.findOrThrow(sdrplay_api_h.java:60)
	at io.github.dsheirer.source.tuner.sdrplay.api.v3_07.sdrplay_api_h$sdrplay_api_Open.<clinit>(sdrplay_api_h.java:1509)
	... 7 more
```

**Additional context**
As reported by @11a11935 on Discord

--- a/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/README
+++ b/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/README
@@ -42,6 +42,7 @@ As a final step, delete the original api version branches and rename each of the
 
 Code Modifications: for each version, modify the sdrplay_api_h.java class to wrap the SYMBOL_LOOKUP variable using the
 following changes, so that the code doesn't throw an error on systems that don't have the API installed:
+
     //jextract auto-generated code modified to wrap with exception handler for systems that don't have the library
     static SymbolLookup SYMBOL_LOOKUP;
     static
@@ -54,9 +55,32 @@ following changes, so that the code doesn't throw an error on systems that don't
         }
         catch(Exception e)
         {
-            SYMBOL_LOOKUP = null;
+            if(SDRPlayLibraryHelper.LOADED_FROM_PATH)
+            {
+                try
+                {
+                    SYMBOL_LOOKUP = SymbolLookup.libraryLookup(SDRPlayLibraryHelper.LIBRARY_PATH, LIBRARY_ARENA)
+                            .or(SymbolLookup.loaderLookup())
+                            .or(Linker.nativeLinker().defaultLookup());
+                }
+                catch(Exception e2)
+                {
+                    SYMBOL_LOOKUP = null;
+                }
+            }
+            else
+            {
+                SYMBOL_LOOKUP = null;
+            }
         }
     }
+
+... and add null checking to the findOrThrow() method ...
+        if(SYMBOL_LOOKUP == null)
+        {
+            throw new UnsatisfiedLinkError("unresolved symbol:" + symbol);
+        }
+
 
 IntelliJ setup
 1. Run configuration

--- a/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/SDRPlayLibraryHelper.java
+++ b/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/SDRPlayLibraryHelper.java
@@ -42,22 +42,22 @@ public class SDRPlayLibraryHelper
     private static final String SDRPLAY_API_PATH_MAC_OS = "/usr/local/lib/libsdrplay_api.dylib";
     private static final String SDRPLAY_API_PATH_MAC_OS_ALTERNATE = "/usr/local/lib/libsdrplay_api.so";
     private static final String SDRPLAY_API_PATH_WINDOWS = System.getenv("ProgramFiles") +
-            "\\SDRplay\\API\\" + (System.getProperty("sun.arch.data.model").contentEquals("64") ? "x64" : "x86") +
-            "\\" + SDRPLAY_API_LIBRARY_NAME;
-    private static final String JAVA_LIBRARY_PATH_KEY = "java.library.path";
-    public static final boolean LOADED = load();
+            "\\SDRplay\\API\\x64\\" + SDRPLAY_API_LIBRARY_NAME;
 
-    /**
-     * Attempts to load the SDRPlay API library from the local system.
-     *
-     * @return true if library was loaded successfully.
-     */
-    public static boolean load()
+    public static final boolean LOADED;
+    public static final boolean LOADED_FROM_PATH;
+    public static final Path LIBRARY_PATH = Path.of(getSDRplayLibraryPath());
+
+    static
     {
+        boolean loaded = false;
+        boolean loadedFromPath = false;
+
         try
         {
             System.loadLibrary(SDRPLAY_API_LIBRARY_NAME);
-            return true;
+            mLog.info("SDRPLay API library loaded by name [" + SDRPLAY_API_LIBRARY_NAME + "]");
+            loaded = true;
         }
         catch(Throwable t)
         {
@@ -70,11 +70,13 @@ public class SDRPlayLibraryHelper
                 try
                 {
                     System.load(libraryPath);
-                    return true;
+                    mLog.info("SDRPLay API library loaded by path [" + libraryPath + "]");
+                    loaded = true;
+                    loadedFromPath = true;
                 }
                 catch(Throwable t2)
                 {
-                    mLog.warn("Unable to load SDRPlay API native library: " + libraryPath);
+                    mLog.info("SDRPlay API native library not found at " + libraryPath);
                 }
             }
             else
@@ -83,7 +85,8 @@ public class SDRPlayLibraryHelper
             }
         }
 
-        return false;
+        LOADED = loaded;
+        LOADED_FROM_PATH = loadedFromPath;
     }
 
     /**
@@ -114,10 +117,5 @@ public class SDRPlayLibraryHelper
 
         mLog.error("Unrecognized operating system.  Cannot identify sdrplay api library path");
         return "";
-    }
-
-    public static void main(String[] args)
-    {
-        System.out.println("Loaded: " + SDRPlayLibraryHelper.LOADED);
     }
 }

--- a/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/v3_07/sdrplay_api_h.java
+++ b/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/v3_07/sdrplay_api_h.java
@@ -21,6 +21,7 @@
 
 package io.github.dsheirer.source.tuner.sdrplay.api.v3_07;
 
+import io.github.dsheirer.source.tuner.sdrplay.api.SDRPlayLibraryHelper;
 import java.lang.foreign.AddressLayout;
 import java.lang.foreign.Arena;
 import java.lang.foreign.FunctionDescriptor;
@@ -57,6 +58,10 @@ public class sdrplay_api_h {
     }
 
     static MemorySegment findOrThrow(String symbol) {
+        if(SYMBOL_LOOKUP == null)
+        {
+            throw new UnsatisfiedLinkError("unresolved symbol:" + symbol);
+        }
         return SYMBOL_LOOKUP.find(symbol)
             .orElseThrow(() -> new UnsatisfiedLinkError("unresolved symbol: " + symbol));
     }
@@ -95,7 +100,23 @@ public class sdrplay_api_h {
         }
         catch(Exception e)
         {
-            SYMBOL_LOOKUP = null;
+            if(SDRPlayLibraryHelper.LOADED_FROM_PATH)
+            {
+                try
+                {
+                    SYMBOL_LOOKUP = SymbolLookup.libraryLookup(SDRPlayLibraryHelper.LIBRARY_PATH, LIBRARY_ARENA)
+                            .or(SymbolLookup.loaderLookup())
+                            .or(Linker.nativeLinker().defaultLookup());
+                }
+                catch(Exception e2)
+                {
+                    SYMBOL_LOOKUP = null;
+                }
+            }
+            else
+            {
+                SYMBOL_LOOKUP = null;
+            }
         }
     }
 

--- a/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/v3_08/sdrplay_api_h.java
+++ b/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/v3_08/sdrplay_api_h.java
@@ -21,6 +21,7 @@
 
 package io.github.dsheirer.source.tuner.sdrplay.api.v3_08;
 
+import io.github.dsheirer.source.tuner.sdrplay.api.SDRPlayLibraryHelper;
 import java.lang.foreign.AddressLayout;
 import java.lang.foreign.Arena;
 import java.lang.foreign.FunctionDescriptor;
@@ -57,6 +58,10 @@ public class sdrplay_api_h {
     }
 
     static MemorySegment findOrThrow(String symbol) {
+        if(SYMBOL_LOOKUP == null)
+        {
+            throw new UnsatisfiedLinkError("unresolved symbol:" + symbol);
+        }
         return SYMBOL_LOOKUP.find(symbol)
             .orElseThrow(() -> new UnsatisfiedLinkError("unresolved symbol: " + symbol));
     }
@@ -95,7 +100,23 @@ public class sdrplay_api_h {
         }
         catch(Exception e)
         {
-            SYMBOL_LOOKUP = null;
+            if(SDRPlayLibraryHelper.LOADED_FROM_PATH)
+            {
+                try
+                {
+                    SYMBOL_LOOKUP = SymbolLookup.libraryLookup(SDRPlayLibraryHelper.LIBRARY_PATH, LIBRARY_ARENA)
+                            .or(SymbolLookup.loaderLookup())
+                            .or(Linker.nativeLinker().defaultLookup());
+                }
+                catch(Exception e2)
+                {
+                    SYMBOL_LOOKUP = null;
+                }
+            }
+            else
+            {
+                SYMBOL_LOOKUP = null;
+            }
         }
     }
 

--- a/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/v3_15/sdrplay_api_h.java
+++ b/src/main/java/io/github/dsheirer/source/tuner/sdrplay/api/v3_15/sdrplay_api_h.java
@@ -21,6 +21,7 @@
 
 package io.github.dsheirer.source.tuner.sdrplay.api.v3_15;
 
+import io.github.dsheirer.source.tuner.sdrplay.api.SDRPlayLibraryHelper;
 import java.lang.foreign.AddressLayout;
 import java.lang.foreign.Arena;
 import java.lang.foreign.FunctionDescriptor;
@@ -57,6 +58,10 @@ public class sdrplay_api_h {
     }
 
     static MemorySegment findOrThrow(String symbol) {
+        if(SYMBOL_LOOKUP == null)
+        {
+            throw new UnsatisfiedLinkError("unresolved symbol:" + symbol);
+        }
         return SYMBOL_LOOKUP.find(symbol)
             .orElseThrow(() -> new UnsatisfiedLinkError("unresolved symbol: " + symbol));
     }
@@ -95,7 +100,23 @@ public class sdrplay_api_h {
         }
         catch(Exception e)
         {
-            SYMBOL_LOOKUP = null;
+            if(SDRPlayLibraryHelper.LOADED_FROM_PATH)
+            {
+                try
+                {
+                    SYMBOL_LOOKUP = SymbolLookup.libraryLookup(SDRPlayLibraryHelper.LIBRARY_PATH, LIBRARY_ARENA)
+                            .or(SymbolLookup.loaderLookup())
+                            .or(Linker.nativeLinker().defaultLookup());
+                }
+                catch(Exception e2)
+                {
+                    SYMBOL_LOOKUP = null;
+                }
+            }
+            else
+            {
+                SYMBOL_LOOKUP = null;
+            }
         }
     }
 

